functions/api/branch.js

```diff
@@ -254,18 +259,51 @@ function validateParams(body, type) {
                     country: joi.string()
                         .min(1)
                         .max(50),
-                    zipcode: joi.number()
+                    zipcode: joi.number().integer().strict()
                 }),
                 isHeadOffice: joi.bool(),
                 isActive: joi.bool(),
                 users: joi.array()
                     .items(joi.string().allow('')),
                 lastUpdatedDate: joi.date(),
-                createdDate: joi.date()
+                createdDate: joi.date(),
+                contact: joi.string(),
+                contactPerson: joi.string()
             })
             break
     }
     return validate(schema, body)
 }
 
-module.exports = router;
\ No newline at end of file
+module.exports = router;
+
+/**
+ * Trigger to execute actions when Branch is deleted
+ */
+module.exports.addOrUpdateBranch = functions.firestore
+    .document(`/${constants.BRANCHES}/{branchId}`)
+    .onWrite(async (change, context) => {
+        const branchId = context.params.branchId
+        if (!change.after._fieldsProto) {
+            logger.debug(`Branch ${change.before.data()[(constants.NAME)]} has been deleted`)
+            deleteThresholdsFromAllProducts(branchId)
+        }
+    });
+
+/**
+ * Method to delete all threshold for deleted branch from all products
+ * @param {*} newProduct
+ */
+async function deleteThresholdsFromAllProducts(branchId) {
+    console.log(`Deleting thresholds for branch ${branchId} from all products`)
+    const productCollecton = db.collection(constants.PRODUCTS)
+    const productSnapshots = await productCollecton.get()
+    productSnapshots.forEach(async product => {
+        const productData = product.data()
+        if(productData[constants.THRESHOLDS].hasOwnProperty(branchId)) {
+            console.log(`Deleting threshold for branch ${branchId} from product ${productData[constants.NAME]}`)
+            delete productData[constants.THRESHOLDS][branchId]
+            await product.ref.update({'thresholds': productData[constants.THRESHOLDS]}, { merge: true })
```

Comment:
i somehow feel, this is an indefinite loop. where you update product and each product keeps coming again in foreach loop. Please double check once.

Reply:
It actually does not. We can sync up and I can explain the flow